examples/user_info_custom_provider/src/main.rs

```diff
@@ -0,0 +1,58 @@
+#![feature(decl_macro, proc_macro_hygiene)]
+
+use rocket::fairing::AdHoc;
+use rocket::http::{Cookie, Cookies};
+use rocket::request::{self, FromRequest, Request};
+use rocket::response::Redirect;
+use rocket::{get, routes, Outcome};
+use rocket_oauth2::OAuthConfig;
+
+mod oauth2;
+
+struct User {
+    pub username: String,
+}
+
+impl<'a, 'r> FromRequest<'a, 'r> for User {
+    type Error = ();
+
+    fn from_request(request: &'a Request<'r>) -> request::Outcome<User, ()> {
+        let mut cookies = request.guard::<Cookies<'_>>().expect("request cookies");
+        if let Some(cookie) = cookies.get_private("username") {
+            return Outcome::Success(User {
+                username: cookie.value().to_string(),
+            });
+        }
+
+        Outcome::Forward(())
+    }
+}
+
+#[get("/")]
+fn index(user: User) -> String {
+    format!("Hi, {}!", user.username)
+}
+
+#[get("/", rank = 2)]
+fn index_anonymous() -> &'static str {
+    "Please login (/login/github)"
+}
+
+#[get("/logout")]
+fn logout(mut cookies: Cookies<'_>) -> Redirect {
+    cookies.remove(Cookie::named("username"));
+    Redirect::to("/")
+}
+
+fn main() {
+    rocket::ignite()
+        .mount("/", routes![index, index_anonymous, logout])
+        .attach(AdHoc::on_attach(
+            "GitHub OAuth2",
+            |rocket| match OAuthConfig::from_config(rocket.config(), "github") {
```

Comment:
Why check `from_config` manually? AFAICT this would suppress a specific message like "Invalid configuration: expected string or table for provider, found integer" leaving only the generic message "A fairing failed: GitHub OAuth2". And in the success case, it would validate the configuration twice for no apparent benefit.

Reply:
Good catch. I believe what happened here is I started building this intending to use the [`OAuth2::custom`](https://docs.rs/rocket_oauth2/0.1.0/rocket_oauth2/struct.OAuth2.html) method and pass in the `OAuthConfig` object directly before realizing that the [`OAuth2::fairing`] method would allow me to simply specify the `config_name: &str` field in the `Rocket.toml`.

But then I forgot to come back and clean up the attachment!